components/style/values/specified/grid.rs

```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
+        // marker to check whether we've already encountered <auto-repeat> along the way
+        let mut is_auto = false;
+        // assume that everything is <fixed-size>. This flag is useful when we encounter <auto-repeat>
+        let mut atleast_one_not_fixed = false;
+
+        loop {
+            current_names = input.try(parse_line_names).unwrap_or(vec![]);
+            if let Ok(track_size) = input.try(|i| TrackSize::parse(context, i)) {
+                if !track_size.is_fixed() {
+                    atleast_one_not_fixed = true;
+                    if is_auto {
+                        return Err(())      // <auto-track-list> only accepts <fixed-size> and <fixed-repeat>
+                    }
+                }
+
+                values.push((current_names, Either::First(track_size)));
+            } else if let Ok((repeat, type_)) = input.try(|i| TrackRepeat::parse_with_repeat_type(context, i)) {
+                if list_type == TrackListType::Explicit {
+                    list_type = TrackListType::Normal;      // <explicit-track-list> doesn't contain repeat()
+                }
+
+                match type_ {
+                    RepeatType::Normal => {
+                        atleast_one_not_fixed = true;
+                        if is_auto {            // only <fixed-repeat>
+                            return Err(())
+                        }
+                    },
+                    RepeatType::Auto => {
+                        if is_auto || atleast_one_not_fixed {
+                            // We've either seen <auto-repeat> earlier, or there's at least one non-fixed value
+                            return Err(())
+                        }
+
+                        is_auto = true;
+                        list_type = TrackListType::Auto(values.len() as u32);
+                    },
+                    RepeatType::Fixed => (),
+                }
+
+                values.push((current_names, Either::Second(repeat)));
+            } else {
+                if values.is_empty() {
+                    return Err(())
+                }
+
+                break
+            }
+        }
+
+        Ok(TrackList {
+            list_type: list_type,
+            values_list: values,
+            line_names_last: current_names,
+        })
+    }
+}
+
+impl<L: ToCss> ToCss for TrackList<L> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        for (i, &(ref names, ref value)) in self.values_list.iter().enumerate() {
+            if i > 0 {
+                dest.write_str(" ")?;
+            }
+
+            if !names.is_empty() {
+                dest.write_str("[")?;
+                dest.write_str(&names.join(" "))?;
+                dest.write_str("] ")?;
+            }
+
+            value.to_css(dest)?;
+        }
+
+        if !self.line_names_last.is_empty() {
+            dest.write_str(" [")?;
+            dest.write_str(&self.line_names_last.join(" "))?;
```

Comment:
ditto. Also, they need to be escaped I believe. Please add a helper function to do so.